File: eco-core/core-plugin/src/main/java/com/willfp/reforges/ReforgesPlugin.java
package com.willfp.reforges;

import com.willfp.eco.core.command.impl.PluginCommand;
import com.willfp.eco.core.display.DisplayModule;
import com.willfp.eco.core.integrations.IntegrationLoader;
import com.willfp.eco.core.items.Items;
import com.willfp.libreforge.LibReforgePlugin;
import com.willfp.reforges.commands.CommandReforge;
import com.willfp.reforges.commands.CommandReforges;
import com.willfp.reforges.config.ReforgesYml;
import com.willfp.reforges.config.TargetYml;
import com.willfp.reforges.display.ReforgesDisplay;
import com.willfp.reforges.integrations.talismans.TalismansIntegration;
import com.willfp.reforges.reforges.Reforges;
import com.willfp.reforges.reforges.util.ReforgeArgParser;
import com.willfp.reforges.reforges.util.ReforgeEnableListeners;
import com.willfp.reforges.reforges.util.ReforgeLookup;
import com.willfp.reforges.util.AntiPlaceListener;
import com.willfp.reforges.util.DiscoverRecipeListener;
import lombok.Getter;
import org.bukkit.event.Listener;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.List;

public class ReforgesPlugin extends LibReforgePlugin {
    /**
     * Instance of Reforges.
     */
    private static ReforgesPlugin instance;

    /**
     * target.yml.
     */
    @Getter
    private final TargetYml targetYml;

    /**
     * reforges.yml.
     */
    @Getter
    private final ReforgesYml reforgesYml;

    /**
     * Internal constructor called by bukkit on plugin load.
     */
    public ReforgesPlugin() {
        super(1330, 12412, "&3", "");
        this.targetYml = new TargetYml(this);
        this.reforgesYml = new ReforgesYml(this);
        instance = this;

        registerHolderProvider(ReforgeLookup::provideReforges);
    }

    @Override
    public void handleEnableAdditional() {
        Items.registerArgParser(new ReforgeArgParser());
    }

    @Override
    public void handleReloadAdditional() {
        this.getLogger().info(Reforges.values().size() + " Reforges Loaded");
    }


    @Override
    protected List<Listener> loadListeners() {
        return Arrays.asList(
                new DiscoverRecipeListener(this),
                new AntiPlaceListener(),
                new ReforgeEnableListeners(this)
        );
    }

    @Override
    protected List<PluginCommand> loadPluginCommands() {
        return Arrays.asList(
                new CommandReforge(this),
                new CommandReforges(this)
        );
    }

    @Override
    protected @Nullable DisplayModule createDisplayModule() {
        return new ReforgesDisplay(this);
    }

    @NotNull
    @Override
    public List<IntegrationLoader> loadAdditionalIntegrations() {
        return Arrays.asList(
                new IntegrationLoader("Talismans", TalismansIntegration::registerProvider)
        );
    }

    @Override
    @NotNull
    public String getMinimumEcoVersion() {
        return "6.35.1";
    }

    /**
     * Get an instance of Reforges.
     *
     * @return The instance.
     */
    public static ReforgesPlugin getInstance() {
        return instance;
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/reforges/ReforgesPlugin.kt
package com.willfp.reforges

import com.willfp.eco.core.command.impl.PluginCommand
import com.willfp.eco.core.display.DisplayModule
import com.willfp.eco.core.items.Items
import com.willfp.libreforge.conditions.Conditions
import com.willfp.libreforge.loader.LibreforgePlugin
import com.willfp.libreforge.loader.configs.ConfigCategory
import com.willfp.libreforge.registerHolderProvider
import com.willfp.reforges.commands.CommandReforge
import com.willfp.reforges.commands.CommandReforges
import com.willfp.reforges.config.TargetYml
import com.willfp.reforges.display.ReforgesDisplay
import com.willfp.reforges.gui.ReforgeGUI
import com.willfp.reforges.libreforge.ConditionHasReforge
import com.willfp.reforges.reforges.PriceMultipliers
import com.willfp.reforges.reforges.ReforgeFinder
import com.willfp.reforges.reforges.ReforgeStoneTag
import com.willfp.reforges.reforges.ReforgeTargets
import com.willfp.reforges.reforges.ReforgedTag
import com.willfp.reforges.reforges.Reforges
import com.willfp.reforges.reforges.util.ReforgeArgParser
import com.willfp.reforges.util.AntiPlaceListener
import com.willfp.reforges.util.DiscoverRecipeListener
import org.bukkit.event.Listener

internal lateinit var plugin: ReforgesPlugin
    private set

class ReforgesPlugin : LibreforgePlugin() {
    val targetYml: TargetYml = TargetYml

    init {
        plugin = this
    }

    override fun loadConfigCategories(): List<ConfigCategory> {
        return listOf(
            Reforges
        )
    }

    override fun handleEnable() {
        Conditions.register(ConditionHasReforge)

        Items.registerArgParser(ReforgeArgParser)

        Items.registerTag(ReforgedTag)
        Items.registerTag(ReforgeStoneTag)

        registerHolderProvider(ReforgeFinder.toHolderProvider())
    }

    override fun handleReload() {
        ReforgeTargets.update()
        PriceMultipliers.update()
        ReforgeGUI.update()
    }

    override fun loadListeners(): List<Listener> {
        return listOf(
            DiscoverRecipeListener,
            AntiPlaceListener,
        )
    }

    override fun loadPluginCommands(): List<PluginCommand> {
        return listOf(
            CommandReforge,
            CommandReforges
        )
    }

    override fun loadDisplayModules(): List<DisplayModule> {
        return listOf(
            ReforgesDisplay
        )
    }
}
